File: GUILib/src/space/kiibou/event/MouseEventListener.java
package space.kiibou.event;

import java.util.Arrays;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.function.Consumer;

public interface MouseEventListener {

    static MouseEventOption options(MouseEventButton button, MouseEventAction action, MouseEventModifier... modifiers) {
        return options(button, EnumSet.of(action), modifiers);
    }

    static MouseEventOption options(MouseEventButton button, EnumSet<MouseEventAction> actions, MouseEventModifier... modifiers) {
        final EnumSet<MouseEventModifier> mods = EnumSet.noneOf(MouseEventModifier.class);
        mods.addAll(Arrays.asList((modifiers)));
        return new MouseEventOption(button, actions, mods);
    }

    default void mouseEvent(MouseEvent event) {
        if (isActive()) {
            MouseOptionMap map = getMouseOptionMap();

            MouseEventOption option = event.getOption();

            map.computeIfPresent(option, (mouseEventOption, mouseEventConsumer) -> {
                mouseEventConsumer.accept(event);
                return mouseEventConsumer;
            });
        }
    }

    boolean isActive();

    void setActive(boolean active);

    default void activate() {
        if (!isActive()) {
            setActive(true);
        }
    }

    default void deactivate() {
        if (isActive()) {
            setActive(false);
        }
    }

    MouseOptionMap getMouseOptionMap();

    default void registerCallback(MouseEventOption option, Consumer<MouseEvent> callback) {
        getMouseOptionMap().merge(option, callback, Consumer::andThen);
    }

    default void unregisterCallback(MouseEventOption option) {
        getMouseOptionMap().remove(option);
    }

    class MouseOptionMap extends HashMap<MouseEventOption, Consumer<MouseEvent>> {
    }

}

File: GUILib/src/space/kiibou/event/MouseEventListener.kt
package space.kiibou.event

import java.util.*

class MouseOptionMap : HashMap<MouseEventOption, (MouseEvent) -> Unit>()
typealias MouseEventConsumer = (MouseEvent) -> Unit

interface MouseEventListener {
    fun mouseEvent(event: MouseEvent) {
        if (active) mouseOptionMap[event.option]?.invoke(event)
    }

    var active: Boolean
    fun activate() {
        if (!active) {
            active = true
        }
    }

    fun deactivate() {
        if (active) {
            active = false
        }
    }

    val mouseOptionMap: MouseOptionMap
    fun registerCallback(option: MouseEventOption, callback: MouseEventConsumer) {
        mouseOptionMap.merge(option, callback) { obj, after -> obj.andThen(after) }
    }

    fun unregisterCallback(option: MouseEventOption) {
        mouseOptionMap.remove(option)
    }
}

inline fun <T> ((T) -> Unit).andThen(crossinline other: (T) -> Unit) = { it: T ->
    this(it)
    other(it)
}

fun options(button: MouseEventButton, action: MouseEventAction, vararg modifiers: MouseEventModifier): MouseEventOption {
    return options(button, EnumSet.of(action), *modifiers)
}

fun options(button: MouseEventButton, actions: EnumSet<MouseEventAction>, vararg modifiers: MouseEventModifier): MouseEventOption {
    val mods = EnumSet.noneOf(MouseEventModifier::class.java)
    mods.addAll(listOf(*modifiers))
    return MouseEventOption(button, actions, mods)
}